### PR TITLE
CP-18348: Production implementation of new patchqueue handling

### PR DESCRIPTION
**Pull request by srowe**
Add the concept of a link file which allows an indirect reference to a spec file and a patchqueue (or set of patches).

Tested building xen SRPM using a modified patchqueue containing the specfile

Testing building dracut SRPM using the CentOS source on gitblit.

### CODE REVIEW

**Comment by euanh** [comment]
You can remove the patch queue handling code from planex-make-srpm now.

File: planex/extract.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+                            check_package_name=check_package_names)
+    urls = [urlparse.urlparse(url) for url in spec.source_urls()]
+    return zip(spec.source_paths(), urls)
+
+
+def extract_file(tar, name_in, name_out):
+    """
+    Extract a file from a tarball
+    """
+    logging.debug("Extracting", name_in, "to", name_out)
```
Comment (euanh) [inline planex/extract.py]: logging.debug("Extracting %s to %s", name_in, name_out)

File: planex/extract.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+import argcomplete
+
+from planex.util import add_common_parser_options
+from planex.util import setup_logging
+from planex.util import setup_sigint_handler
+import planex.spec
+
+
+def all_sources(spec, topdir, check_package_names):
```
Comment (euanh) [inline planex/extract.py]: Please move this into the Spec class.   You can then factor it out of fetch.py as well.

File: planex/extract.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    return zip(spec.source_paths(), urls)
+
+
+def extract_file(tar, name_in, name_out):
+    """
+    Extract a file from a tarball
+    """
+    logging.debug("Extracting", name_in, "to", name_out)
+    if name_in not in tar.getnames():
+        sys.exit("%s: %s not found" % (sys.argv[0], name_in))
```
Comment (euanh) [inline planex/extract.py]: "not found" -> "not found in archive"

File: planex/extract.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+    spec = planex.spec.Spec(args.output, topdir=args.topdir,
+                            check_package_name=args.check_package_names)
+    patchnum = spec.highest_patch()
+
+    # Rewrite the spec file to include the patches
+    rewrite_spec(args.output, args.output+".new", patches, patchnum)
+
+    # Switch extracted spec file with new specfile
+    os.rename(args.output, args.output+".old")
```
Comment (euanh) [inline planex/extract.py]: Need spaces around + operator.

File: planex/extract.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+        with open(args.link) as fileh:
+            link = json.load(fileh)
+
+    except IOError as exn:
+        # IO error loading JSON file
+        sys.exit("%s: %s: %s" %
+                 (sys.argv[0], exn.strerror, exn.filename))
+
+    # Extract the spec file
+    tar = tarfile.open(args.tarball)
```
Comment (euanh) [inline planex/extract.py]: The Tarfile object can be used as a context manager.   Please change this to 'with tarfile.open(...) as tar'

Comment (srowe) [inline planex/extract.py]: I'll try that again but it barfed when I did before.

File: planex/extract.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+def extract_file(tar, name_in, name_out):
+    """
+    Extract a file from a tarball
+    """
+    logging.debug("Extracting %s to %s", name_in, name_out)
+    if name_in not in tar.getnames():
+        sys.exit("%s: %s not found" % (sys.argv[0], name_in))
+    try:
+        fh_in = tar.extractfile(name_in)
```
Comment (euanh) [inline planex/extract.py]: try
```
from contextlib import closing
...
with closing(tar.extractfile(name_in)) as f:
   ...
```

File: planex/extract.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+                fh_out.write(line)
+
+
+def expand_patchqueue(args, tar, seriesfile):
+    """
+    Create a list of patches from a patchqueue and update the spec file
+    """
+    # Build a list of patches
+    try:
+        series = tar.extractfile(seriesfile)
```
Comment (euanh) [inline planex/extract.py]: try
```
from contextlib import closing
...
with closing(tar.extractfile(seriesfile)) as series:
   patches = list(parse_patchseries(series))
```

File: planex/extract.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    """
+    Parse series file and return the list of patches
+    """
+    guard_re = re.compile(r'([\S]+)(\s#.*)?')
+
+    for line in series:
+        line = line.strip()
+        if not line or line.startswith('#'):
+            continue
+
```
Comment (euanh) [inline planex/extract.py]: Please change the default value of guard from '' to None, add `if guard is None: continue` here.

Comment (srowe) [inline planex/extract.py]: That won't have the right behaviour with negative patchguards. They need to fall through to the yield().

**Comment by srowe** [comment]
The only failure now is a complaint the main() boilerplate code is the same in pins and extract. How do we fix that?

**Pull request merged**